crates/core/src/kakarot/mod.rs

```diff
@@ -0,0 +1,123 @@
+use std::marker::PhantomData;
+
+use starknet::core::types::{BlockId, FunctionCall};
+use starknet::providers::Provider;
+use starknet_crypto::FieldElement;
+
+use crate::client::constants::selectors::{BALANCE_OF, BYTECODE, COMPUTE_STARKNET_ADDRESS, ETH_CALL, GET_EVM_ADDRESS};
+use crate::client::constants::STARKNET_NATIVE_TOKEN;
+use crate::client::errors::EthApiError;
+use crate::client::helpers::DataDecodingError;
+
+pub struct KakarotContract<P> {
+    pub address: FieldElement,
+    pub proxy_account_class_hash: FieldElement,
+    _phantom: PhantomData<P>,
+}
+
+impl<P: Provider + Send + Sync> KakarotContract<P> {
+    #[must_use]
+    pub fn new(address: FieldElement, proxy_account_class_hash: FieldElement) -> Self {
+        Self { address, proxy_account_class_hash, _phantom: PhantomData }
+    }
+
+    pub async fn balance(
```

Comment:
wdyt about creating a `EthContract` on the same model for this `balanceOf` ?

Reply:
Should I do `EthContract` or `Erc20Contract`?

Reply:
`Erc20Contract` indeed